think-dock.1.rst
##########
think-dock
##########

**********************************************************
set the screens when going to and from the docking station
**********************************************************

:Author: Martin Ueding <dev@martin-ueding.de>
:Date: 2012-02-27
:Manual section: 1

SYNOPSIS
========

::

    think-dock [on|off]

DESCRIPTION
===========

This program sets the screen resolution correctly when putting the ThinkPad
onto the docking station. It also sets the Wacom input devices to act on the
internal screen only.

It deduces what to do automatically, if no option is given. If it is docked, it
will perform the docking action. When you pressed the eject button on the
docking station, it will un-dock.

There will be an udev rule installed that will automatically dock it when set
onto the station and un-dock when you press the eject button.

what it does
------------

When docking, the following things are done:

- Activating the external monitor.
- Setting the external monitor as primary monitor.
- Deactivate the wireless connection.
- Set the Wacom devices to the internal screen only.
- Set the brightness to a fixed value, currently 60%.
- Unmute the speakers and set the volume to 100%.

When undocking, the following things are done:

- Deactivating external monitor.
- Setting the internal monitor as primary monitor.
- Activating the wireless connection.
- Set the speakers to some medium volume, currently 50%.

OPTIONS
=======

on|off
    If you have it sitting on the docking station and want it to dock, use
    ``on``. Otherwise use ``off`` before you take the ThinkPad off the docking
    station.

    You can omit this option and the script will guess what to do.

EXIT STATUS
===========

0
    Everything okay.
1
    Some error.

FILES
=====

You can create a config file in ``$HOME/.config/think-rotate/dock.sh``, which
is a simple Bash script that is going to be sourced from ``think-dock``.

A sample config would look like this::

    dock_loudness="50%"
    disable_wifi=false
    relative_position=left

You can set the following options:

``disable_wifi``
    Whether to set the wifi. Set it to ``true`` or something else.

``internal``
    The ``xrandr`` name for the internal monitor.

``unmute``
    Whether to change the volume. Set it to ``true`` or something else.

``dock_loudness``
    Volume to set to when docking. Set it to a percentage like ``100%``.

``undock_loundess``
    Volume to set to when undocking. Set it to a percentage like ``50%``.

``set_brightness``
    Whether to change the brightness. Set it to ``true`` or something else.

``brightness``
    Brightness to set to when docking. Set it to a percentage like ``60%``.

``relative_position``
    Where to set the external monitor. Set it to ``right`` or ``left`` or
    anything else that ``xrandr`` supports with a ``--*-of`` argument.

EXAMPLE
=======

You can just call ``think-dock`` and it will do the right think probably.

If you want, you can tell the script what to do: When you have it sitting on
the docking station, call ``think-dock on`` to get the external screen going.
When you are done, call ``think-dock off`` before you disconnect to get the
internal screen back again.


CHANGELOG
.. Copyright © 2012 Martin Ueding <dev@martin-ueding.de>

#########
changelog
#########

v2.3
    - Add support for other virtual keyboards. Thanks to Jim Turner!
    - Use shorter redirection (``&>`` instead of ``2>&!``).

v2.2.1
    - Fix spelling typo in ``relative_position``. Thanks to Jim Turner!

v2.2
    - Background most tasks so that they run in parallel. This should speed up
      docking.

v2.1
    - Only set Wacom screen devices. That way, any attached Wacom graphics
      tablet is not affected by the docking.

v2.0
    - Use the kernel to determine what the docking status is.
    - Add ``udev`` rules to perform the docking action.

v1.5
    - Desktop files for think-dock.

v1.4.3
    - Disable the wireless connection on docking.

v1.4.2
    - Fix commands in ``.desktop`` files.

v1.4.1
    - Install ``.desktop`` files.

v1.4
    - Query the state of the whole system automatically and determine the right
      action. You can still specify ``on`` or ``off``, if you want to.

v1.3
    - Optional config file for ``think-dock``.

v1.2.2
    - Fix flip direction.

v1.2.1
    - Disable wireless only when eth0 connected.
    - Document options.

v1.2
    - Change display brightness on docking.

v1.1
    - Check whether programs are there before using them.
    - Create directories on ``make install``.
    - Disable wifi when going onto the docking station.
    - Enable sound on docking.
    - Lower the volume after docking.
    - Query Wacom devices automatically.

v1.0
    This is the first release with a version number. It contains a couple fixes
    and improvements compared to previous (before 2012-07-13) versions of these
    scripts.

    - Accept other names for the rotation.
    - Disable the trackpad as well.
    - Start and stop the virtual keyboard.
    - Try to go back automatically, if a rotation is already set.
    - Use ``--rotation`` instead of ``-o``. This will only rotate the internal
      screen and not any attached screens as well.

Way before 2012-07-13, those are significant changes in the history:

- Add desktop files.
- Also set Wacom hardware correctly.
- Determine resolution automatically.
- Disable trackpoint when switching.
- Dynamically find external display.
- Limit Wacom devices to internal screen.
- Set external monitor as primary.

.. vim: ft=rst


think-dock.py
#!/usr/bin/python
# -*- coding: utf-8 -*-

# Copyright (c) 2012 Martin Ueding <dev@martin-ueding.de>

# This program is free software: you can redistribute it and/or modify it
# under the terms of the GNU General Public License as published by the Free
# Software Foundation, either version 2 of the License, or (at your option)
# any later version.

# This program is distributed in the hope that it will be useful, but WITHOUT
# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
# FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for
# more details.

# You should have received a copy of the GNU General Public License along with
# this program. If not, see <http://www.gnu.org/licenses/>.

import argparse
import subprocess
import re

__docformat__ = "restructuredtext en"

def main():
    internal = "LVDS1"

    options = _parse_args()

    external, is_enabled = find_external(internal)

    print external, is_enabled

    if is_enabled:
		subprocess.check_call(["xrandr", "--output", internal, "--auto", "--primary"])
		subprocess.check_call(["xrandr", "--output", external, "--off"])
    else:
		subprocess.check_call(["xrandr", "--output", internal, "--auto"])
		subprocess.check_call(["xrandr", "--output", external, "--auto", "--right-of", internal, "--primary"])

    for device in ["Wacom ISDv4 E6 Pen stylus", "Wacom ISDv4 E6 Finger touch", "Wacom ISDv4 E6 Pen eraser"]:
        subprocess.check_call(["xsetwacom", "set", device, "MapToOutput", internal])

def find_external(internal):
    """
    Finds the port which has the external monitor attached.

    :param internal: Name of internal monitor, e. g. ``LVDS1``.
    :type internal: str
    :return: Name of the interface and whether the screen is enabled.
    :rtype: Tuple
    """
    xrandr = subprocess.check_output(["xrandr"]).split("\n")

    regex = re.compile(r"^(\S+) (disconnected|connected).*$")

    word_list = []
    interfaces = []

    for line in xrandr:
        words = line.split()
        word_list.append(words)

    for lineno in range(len(word_list)):
        if len(word_list[lineno]) < 2:
            continue

        if word_list[lineno][1] == 'connected':
            port = word_list[lineno][0]
            if port == internal:
                continue

            selected_frequency = word_list[lineno+1][1]

            is_enabled = "*" in selected_frequency

            return port, is_enabled

    return None, None

def _parse_args():
    """
    Parses the command line arguments.

    :return: Namespace with arguments.
    :rtype: Namespace
    """
    parser = argparse.ArgumentParser(usage="", description="")
    #parser.add_argument("action", metavar="action", type=str, nargs=1, help="on|off")
    #parser.add_argument("", dest="", type="", default=, help=)
    #parser.add_argument("--version", action="version", version="<the version>")

    return parser.parse_args()


if __name__ == "__main__":
    main()
